implement the Md5Provider to conform to the HashProvider interface

KarimElghamry opened the pull request:
* implement the the Md5Provider class that implements the HashProvider
  interface
* add the Md5Provider to HashProvider factory
* implement Md5Provider unit tests

Fix #17

Signed-off-by: Karim Elghamry <karimelghamry@gmail.com>

tsconn23 commented on src/main/java/com/alvarium/hash/Md5Provider.java:
```diff
@@ -0,0 +1,24 @@
+class Md5Provider implements HashProvider{
+  private final MessageDigest md5;
+
+  public Md5Provider() throws HashTypeException{
```
Make this (and the other provider constructors) `protected` so that classes outside of this package can't arbitrarily instantiate the classes directly.

KarimElghamry replied:
fixed the md5 provider constructor visibility in 77f26aa0c8377b51ccf22dacbc86372d9dc7ceb9.

tsconn23 commented:
@KarimElghamry Please rebase and resolve conflicts when you get a chance, then update the PR branch

KarimElghamry commented:
> @KarimElghamry Please rebase and resolve conflicts when you get a chance, then update the PR branch

@tsconn23 I've rebased and fixed merge conflicts in 46cffe818fca66e0bd2e55234fc996b54383c736.
